File: AI/ai.cpp
#include "ai.h"
#include "logic.h"
#include "MainDarkLogic/action.h"
#include "MainDarkLogic/EnumFun.h"
#include "masteraithread.h"
#include <chrono>
#include <condition_variable>
#include <iostream>

AI::AI(const size_t& maxInstanceIdx, const size_t& secondTimeout) : Player("BasicAI", 0, 1955),
m_secondTimeout(secondTimeout), m_masterThread(std::make_shared<MasterAIThread>(maxInstanceIdx,*this)), 
	m_crtNode(std::make_unique<Node>(*this)), m_hasEvents(false), m_lock(m_mutex)
{
	m_masterThread->init();
}

AI::AI(const std::string& name_, const size_t& maxInstanceIdx, const size_t timeoutScds) : Player(name_, 0, 1955),
m_secondTimeout(timeoutScds), m_masterThread(std::make_shared<MasterAIThread>(maxInstanceIdx, *this)),
m_crtNode(std::make_unique<Node>(*this)), m_hasEvents(false), m_lock(m_mutex)
{
	m_masterThread->init();
}

std::shared_ptr<const Action> AI::play()
{	
	//start chrono with cadence=1s
	std::chrono::high_resolution_clock::time_point start, end;
	start = std::chrono::high_resolution_clock::now();
	Node* newNode = m_crtNode->getDemoMode();
	std::unique_ptr<size_t> nbNode = nullptr;
	
	//if AI has not found demonstration yet
	if (!newNode)
	{
		m_masterThread->start();
		std::cv_status status = std::cv_status::no_timeout;
		while (!m_hasEvents && (status!=std::cv_status::timeout))
		{
			status = m_condition_var.wait_until(m_lock, start + std::chrono::seconds(m_secondTimeout)); 			
		}
		if (status == std::cv_status::timeout)
		{
			m_masterThread->stop();
		}

		//waiting for threads to stop
		while (!m_hasEvents)
		{
			m_condition_var.wait(m_lock);
		}
		if (m_eventQueue.size())
		{
			//Consuming event
			m_eventQueue.pop();
		}
		m_hasEvents = false;

		nbNode = std::make_unique<size_t>(m_crtNode->nbNode());
		newNode = m_crtNode->getBestNode();
		std::cout<< "[DEBUG] Nb explored node " << *nbNode << std::endl;
	}
	else
	{
		m_masterThread->computeActions();
	}
	
	m_crtNode.reset(newNode);
	m_crtNode->decrDepth();
	m_masterThread->updateLogic(m_crtNode->actionId());
	end = std::chrono::high_resolution_clock::now();
	double elapsed_seconds = std::chrono::duration<double>(end - start).count();
	
	if (nbNode)
	{
		std::cout << "[DEBUG] AI play action " << m_crtNode->actionId() << " with value " << value()
			<< " in " << elapsed_seconds << " seconds" << std::endl;
		std::cout << "[DEBUG] AI exploration speed: " << (*nbNode) / elapsed_seconds << " nodes/s" << std::endl;
	}
		
	return std::make_shared<const Action>(Fun::PUSH_ACTION, m_crtNode->actionId());
}

void AI::meditate()
{
	Player::meditate();
	m_crtNode = std::make_unique<Node>(*this);
}

void AI::pushCrtAction(const Id& actionId, const unsigned char threadIdx) const
{
	m_crtNode->pushCrtAction(actionId, threadIdx);
}

void AI::stopFromMasterThread()
{
	_pushEvent(Event::EventEnum::STOP);
}

void AI::explore(DbAction& dbActions, unsigned char /*threadIdx*/)
{
	m_crtNode->exploreBasic(dbActions.actions());
}

unsigned short int AI::eval(const std::vector<N_DarkLogic::State>&, const unsigned char)
{
	return 0;
}

bool AI::mustStop(const unsigned char threadIdx) const
{
	return m_masterThread->mustStop(threadIdx);
}

void AI::stopThread(const unsigned char threadIdx) const
{
	m_masterThread->stopThread(threadIdx);
}

std::shared_ptr<MasterAIThread> AI::getMaster() const
{
	return m_masterThread;
}

size_t AI::timeout() const
{
	return m_secondTimeout;
}

unsigned int AI::value() const
{
	return m_crtNode->value();
}

unsigned short int AI::getValueFromAction(const unsigned short int action)
{
	return m_crtNode->getValueFromAction(action);
}

unsigned short int AI::getRealValueFromAction(const unsigned short int action)
{
	return m_crtNode->getRealValueFromAction(action);
}

void AI::_pushEvent(Event::EventEnum type_)
{
	m_mutex.lock();
	m_eventQueue.push(Event(0, type_));
	m_mutex.unlock();
	m_hasEvents = true;
	m_condition_var.notify_all();
}

File: PyDarkLogic/AI/ai.py
import MainDarkLogic.darklogic as DarkLogic
from MainDarkLogic.player import Player
from AI.masteraithread import MasterAIThread
from AI.node import Node
from threading import Thread, Condition, Lock
import time
from AI.event import Event
from MainDarkLogic.action import Action
from MainDarkLogic.enumfun import EnumFun


class AI(Player):
    def __init__(self, logicGame, maxInstanceIdx, secondTimeout, name="AI"):
        super().__init__(logicGame, name)
        self._secondTimeout = secondTimeout
        self._maxInstanceIdx = maxInstanceIdx
        self._masterThread = None
        # self._masterThread = MasterAIThread(maxInstanceIdx, self)
        self._crtNode = Node(ai=self)
        self._hasEvents = False
        self._eventQueue = []
        self._mutex = Lock()
        self._condition_var = Condition(self._mutex)
        self._elo = 1700

    def play(self):
        # start chrono
        start = time.perf_counter()
        newNode = self._crtNode.getDemoMode()
        nbNode = None
        hasTimedout = False
        if not newNode:
            self._masterThread = MasterAIThread(self._maxInstanceIdx, self)
            self._masterThread.start_()
            while not self._hasEvents:
                with self._condition_var:
                    hasTimedout = not self._condition_var.wait(self._secondTimeout)
                    if hasTimedout:
                        break

            if hasTimedout:
                self._masterThread.stop_()
            # waiting for threads to stop
            while not self._hasEvents:
                with self._condition_var:
                    self._condition_var.wait()
            if len(self._eventQueue):
                self._eventQueue.pop()
            self._hasEvents = False

            nbNode = self._crtNode.nbNode()
            newNode = self.getBestNode()
            print("[DEBUG] Nb explored nodes: " + str(nbNode))

        else:
            self._masterThread.computeActions()

        self._storeCrtNode()
        self._crtNode = newNode
        self._crtNode.setRoot()
        self._masterThread.updateLogic(self._crtNode.actionId())
        end = time.perf_counter()
        elapsed_seconds = end - start
        if nbNode:
            print("[DEBUG] AI play action " + str(self._crtNode.actionId()) + " with value " +
                  str(self.value()) +
                  " in " + str(elapsed_seconds) + " seconds")
            print("[DEBUG] AI exploration speed: " + str(nbNode / elapsed_seconds) + " nodes/s")

        return Action(EnumFun.PUSH_ACTION, self._crtNode.actionId())

    def pushCrtAction(self, actionIds, threadIdx=-1):
        self._crtNode.pushCrtAction(actionIds, threadIdx)

    def stopFromMasterThread(self):
        self._pushEvent(Event.EventEnum.STOP)

    def setSecondTimeout(self, secondTimeout):
        self._secondTimeout = secondTimeout

    def explore(self, dbNode, threadId):
        self._crtNode.exploreStatic(dbNode.actions(), threadId)

    def mustStop(self, threadIdx):
        return self._masterThread.mustStop(threadIdx)

    def stopThread(self, threadIdx):
        self._masterThread.stopThread(threadIdx)

    def getMaster(self):
        return self._masterThread

    def timeout(self):
        return self._secondTimeout

    def _pushEvent(self, type_):
        self._mutex.acquire()
        self._eventQueue.append(Event(0, type_))
        self._mutex.release()
        self._hasEvents = True
        with self._condition_var:
            self._condition_var.notify_all()

    def _storeCrtNode(self):
        pass

    def meditate(self):
        self._crtNode = Node(ai=self)

    def getBestNode(self):
        return self._crtNode.getBestNode()

    def value(self):
        return self._crtNode.value()

    def eval(self, states, threadId):
        return 0

    def valueOfActions(self, actions):
        return self._crtNode.getValueOfActions(actions)

    def realValueOfActions(self, actions):
        return self._crtNode.getRealValueOfActions(actions)
